README.md
![Static Analysis](https://github.com/nasa/tblCRCTool/workflows/Static%20Analysis/badge.svg)
![Format Check](https://github.com/nasa/tblCRCTool/workflows/Format%20Check/badge.svg)

# Core Flight System : Framework : Tool : Table CRC Generator

This repository contains NASA's Table CRC Generator Tool (tblCRCTool), which is a framework component of the Core Flight System.

This lab application is a ground utility to generate binary table CRCs for cFS. It is intended to be located in the `tools/tblCRCTool` subdirectory of a cFS Mission Tree.  The Core Flight System is bundled at https://github.com/nasa/cFS (which includes this tool as a submodule), which includes build and execution instructions.

## Version Notes

### Development Build: v1.3.0-rc4+dev18
- https://github.com/nasa/tblCRCTool/pull/70
- See <https://github.com/nasa/tblCRCTool/pull/70>

### Development Build: v1.3.0-rc4+dev14

- Update Copyright Headers
- Standardize version.h 
- See <https://github.com/nasa/osal/pull/67> and <https://github.com/nasa/cFS/445>

### Development Build: v1.3.0-rc4+dev7

- Remove explicit filename doxygen comments
- See <https://github.com/nasa/cFS/pull/432>

### Development Build: v1.3.0-rc4+dev2

- Set baseline for cFS-Caelum-rc4: v1.3.0-rc4
- Update mission rev to use 0xFF for development version
- See <https://github.com/nasa/cfs/pull/390>

### Development Build: v1.2.0-rc1+dev33

- Add printf conversion casts to fix Raspberry-Pi4 compilation error
  - Cast ssize_t to long and uint32 to unsigned long for printf.
  - This matches the %d and %lX conversions, respectively.
- Implement Coding Standard in CodeQL
- See <https://github.com/nasa/tblCRCTool/pull/52> and <https://github.com/nasa/cFS/pull/270>

### Development Build: v1.2.0-rc1+dev25

- Add Testing Tools to the Security Policy
- Check lseek return and exit/error processing updates
- See <https://github.com/nasa/tblCRCTool/pull/89>

### Development Build: v1.2.0-rc1+dev19

- Changes CLI "help" option to use two dashes: `--help`
- Adds static analysis and format check to continuous integration workflow. Adds workflow status badges to ReadMe.
- Adds CodeQL Analysis to continuous integration workflow.
- See <https://github.com/nasa/tblCRCTool/pull/35>

### Development Build: v1.2.0-rc1+dev12

- Documentation: Add `Security.md` with instructions on reporting vulnerabilities
- Removes unimplemented CRC cases to eliminate static analysis warnings
- See <https://github.com/nasa/tblCRCTool/pull/29>

### Development Build: v1.2.0-rc1+dev3

- Use `sizeof()` instead of a hard coded value for the table file header size to keep this tool in sync if the size of the cFE file or table header should ever change.
- Update version baseline to v1.2.0-rc1
- Set REVISION number to 99 to indicate development version
- See <https://github.com/nasa/tblCRCTool/pull/25>

### Development Build: 1.1.0+dev7

- Create a version header file
- Report version when responding to `-help` command
- See <https://github.com/nasa/tblCRCTool/pull/22>

### Development Build: 1.1.1

- Apply Code Style
- See <https://github.com/nasa/tblCRCTool/pull/18>

### **_OFFICIAL RELEASE: 1.1.0 - Aquila_**

- Minor updates (see https://github.com/nasa/tblCRCTool/pull/12)
- Released as part of cFE 6.7.0, Apache 2.0

### **_OFFICIAL RELEASE: 1.0.0a_**

- Released as part of cFE 6.6.0a, Apache 2.0

NOTE - there are other parameter set management schemes used with the cFS (JSON, csv, etc) which may be more applicable for modern missions.  Contact the community as detailed below for more information.

## Known issues

This ground utility was developed for a specific mission/configuration, and may not be applicable for general use.

## Getting Help

For best results, submit issues:questions or issues:help wanted requests at https://github.com/nasa/cFS.

Official cFS page: http://cfs.gsfc.nasa.gov


CHANGELOG.md
# Changelog

## Development Build: v1.3.0+rc4+dev24
- Create CHANGELOG.md
- See <https://github.com/nasa/tblCRCTool/pull/73>

## Development Build: v1.3.0-rc4+dev18
- https://github.com/nasa/tblCRCTool/pull/70
- See <https://github.com/nasa/tblCRCTool/pull/70>

## Development Build: v1.3.0-rc4+dev14
- Update Copyright Headers
- Standardize version.h 
- See <https://github.com/nasa/osal/pull/67> and <https://github.com/nasa/cFS/445>

## Development Build: v1.3.0-rc4+dev7
- Remove explicit filename doxygen comments
- See <https://github.com/nasa/cFS/pull/432>

## Development Build: v1.3.0-rc4+dev2
- Set baseline for cFS-Caelum-rc4: v1.3.0-rc4
- Update mission rev to use 0xFF for development version
- See <https://github.com/nasa/cfs/pull/390>

## Development Build: v1.2.0-rc1+dev33
- Add printf conversion casts to fix Raspberry-Pi4 compilation error
  - Cast ssize_t to long and uint32 to unsigned long for printf.
  - This matches the %d and %lX conversions, respectively.
- Implement Coding Standard in CodeQL
- See <https://github.com/nasa/tblCRCTool/pull/52> and <https://github.com/nasa/cFS/pull/270>

## Development Build: v1.2.0-rc1+dev25
- Add Testing Tools to the Security Policy
- Check lseek return and exit/error processing updates
- See <https://github.com/nasa/tblCRCTool/pull/89>

## Development Build: v1.2.0-rc1+dev19
- Changes CLI "help" option to use two dashes: `--help`
- Adds static analysis and format check to continuous integration workflow. Adds workflow status badges to ReadMe.
- Adds CodeQL Analysis to continuous integration workflow.
- See <https://github.com/nasa/tblCRCTool/pull/35>

## Development Build: v1.2.0-rc1+dev12
- Documentation: Add `Security.md` with instructions on reporting vulnerabilities
- Removes unimplemented CRC cases to eliminate static analysis warnings
- See <https://github.com/nasa/tblCRCTool/pull/29>

## Development Build: v1.2.0-rc1+dev3
- Use `sizeof()` instead of a hard coded value for the table file header size to keep this tool in sync if the size of the cFE file or table header should ever change.
- Update version baseline to v1.2.0-rc1
- Set REVISION number to 99 to indicate development version
- See <https://github.com/nasa/tblCRCTool/pull/25>

## Development Build: 1.1.0+dev7
- Create a version header file
- Report version when responding to `-help` command
- See <https://github.com/nasa/tblCRCTool/pull/22>

## Development Build: 1.1.1
- Apply Code Style
- See <https://github.com/nasa/tblCRCTool/pull/18>

## **_OFFICIAL RELEASE: 1.1.0 - Aquila_**
- Minor updates (see https://github.com/nasa/tblCRCTool/pull/12)
- Released as part of cFE 6.7.0, Apache 2.0

## **_OFFICIAL RELEASE: 1.0.0a_**
- Released as part of cFE 6.6.0a, Apache 2.0

NOTE - there are other parameter set management schemes used with the cFS (JSON, csv, etc) which may be more applicable for modern missions.  Contact the community as detailed below for more information.

## Known issues

This ground utility was developed for a specific mission/configuration, and may not be applicable for general use.

## Getting Help

For best results, submit issues:questions or issues:help wanted requests at https://github.com/nasa/cFS.

Official cFS page: http://cfs.gsfc.nasa.gov


cfe_ts_crc.c
/************************************************************************
 * NASA Docket No. GSC-18,719-1, and identified as “core Flight System: Bootes”
 *
 * Copyright (c) 2020 United States Government as represented by the
 * Administrator of the National Aeronautics and Space Administration.
 * All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ************************************************************************/

/*
 *  This program calculates the CRC-16/ARC of a given table file.
 *
 *  Algorithm:
 *    - Name: CRC-16/ARC
 *    - Polynomial: 0x8005
 *    - Initialization: 0x0000
 *    - Reflect Input/Output: true
 *    - XorOut: 0x0000
 *
 *  Inputs: One string containing the filename of the table file to CRC.
 *
 *  Outputs: Prints to the terminal the filename, size, and CRC.
 *           Returns 0 if successful.
 *
 *  Author: Mike Blau, GSFC Code 582
 */
#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>

#include "cfe_ts_crc_version.h"

/* These headers are needed for CFE_FS_Header_t and CFE_TBL_File_Hdr_t, respectively.
 * This uses the OSAL definition of fixed-width types, even thought this tool
 * is not using OSAL itself.
 */
#include "common_types.h"
#include "cfe_fs_extern_typedefs.h"
#include "cfe_tbl_extern_typedefs.h"

/*
**             Function Prologue
**
** Function: CalculateCRC  (originated from lro-cfe-4.2.1 delivery - 2/4/09)
**
** Purpose:  Perform a CRC calculation on a range of memory.
**
*/
uint32 CalculateCRC(void *DataPtr, uint32 DataLength, uint32 InputCRC)
{
    int32  i;
    int16  Index;
    int16  Crc = 0;
    uint8 *BufPtr;

    static const uint16 CrcTable[256] = {

        0x0000, 0xC0C1, 0xC181, 0x0140, 0xC301, 0x03C0, 0x0280, 0xC241, 0xC601, 0x06C0, 0x0780, 0xC741, 0x0500, 0xC5C1,
        0xC481, 0x0440, 0xCC01, 0x0CC0, 0x0D80, 0xCD41, 0x0F00, 0xCFC1, 0xCE81, 0x0E40, 0x0A00, 0xCAC1, 0xCB81, 0x0B40,
        0xC901, 0x09C0, 0x0880, 0xC841, 0xD801, 0x18C0, 0x1980, 0xD941, 0x1B00, 0xDBC1, 0xDA81, 0x1A40, 0x1E00, 0xDEC1,
        0xDF81, 0x1F40, 0xDD01, 0x1DC0, 0x1C80, 0xDC41, 0x1400, 0xD4C1, 0xD581, 0x1540, 0xD701, 0x17C0, 0x1680, 0xD641,
        0xD201, 0x12C0, 0x1380, 0xD341, 0x1100, 0xD1C1, 0xD081, 0x1040, 0xF001, 0x30C0, 0x3180, 0xF141, 0x3300, 0xF3C1,
        0xF281, 0x3240, 0x3600, 0xF6C1, 0xF781, 0x3740, 0xF501, 0x35C0, 0x3480, 0xF441, 0x3C00, 0xFCC1, 0xFD81, 0x3D40,
        0xFF01, 0x3FC0, 0x3E80, 0xFE41, 0xFA01, 0x3AC0, 0x3B80, 0xFB41, 0x3900, 0xF9C1, 0xF881, 0x3840, 0x2800, 0xE8C1,
        0xE981, 0x2940, 0xEB01, 0x2BC0, 0x2A80, 0xEA41, 0xEE01, 0x2EC0, 0x2F80, 0xEF41, 0x2D00, 0xEDC1, 0xEC81, 0x2C40,
        0xE401, 0x24C0, 0x2580, 0xE541, 0x2700, 0xE7C1, 0xE681, 0x2640, 0x2200, 0xE2C1, 0xE381, 0x2340, 0xE101, 0x21C0,
        0x2080, 0xE041, 0xA001, 0x60C0, 0x6180, 0xA141, 0x6300, 0xA3C1, 0xA281, 0x6240, 0x6600, 0xA6C1, 0xA781, 0x6740,
        0xA501, 0x65C0, 0x6480, 0xA441, 0x6C00, 0xACC1, 0xAD81, 0x6D40, 0xAF01, 0x6FC0, 0x6E80, 0xAE41, 0xAA01, 0x6AC0,
        0x6B80, 0xAB41, 0x6900, 0xA9C1, 0xA881, 0x6840, 0x7800, 0xB8C1, 0xB981, 0x7940, 0xBB01, 0x7BC0, 0x7A80, 0xBA41,
        0xBE01, 0x7EC0, 0x7F80, 0xBF41, 0x7D00, 0xBDC1, 0xBC81, 0x7C40, 0xB401, 0x74C0, 0x7580, 0xB541, 0x7700, 0xB7C1,
        0xB681, 0x7640, 0x7200, 0xB2C1, 0xB381, 0x7340, 0xB101, 0x71C0, 0x7080, 0xB041, 0x5000, 0x90C1, 0x9181, 0x5140,
        0x9301, 0x53C0, 0x5280, 0x9241, 0x9601, 0x56C0, 0x5780, 0x9741, 0x5500, 0x95C1, 0x9481, 0x5440, 0x9C01, 0x5CC0,
        0x5D80, 0x9D41, 0x5F00, 0x9FC1, 0x9E81, 0x5E40, 0x5A00, 0x9AC1, 0x9B81, 0x5B40, 0x9901, 0x59C0, 0x5880, 0x9841,
        0x8801, 0x48C0, 0x4980, 0x8941, 0x4B00, 0x8BC1, 0x8A81, 0x4A40, 0x4E00, 0x8EC1, 0x8F81, 0x4F40, 0x8D01, 0x4DC0,
        0x4C80, 0x8C41, 0x4400, 0x84C1, 0x8581, 0x4540, 0x8701, 0x47C0, 0x4680, 0x8641, 0x8201, 0x42C0, 0x4380, 0x8341,
        0x4100, 0x81C1, 0x8081, 0x4040};

    Crc    = (int16)(0xFFFF & InputCRC);
    BufPtr = (uint8 *)DataPtr;

    for (i = 0; i < DataLength; i++, BufPtr++)
    {
        Index = ((Crc ^ *BufPtr) & 0x00FF);
        Crc   = ((Crc >> 8) & 0x00FF) ^ CrcTable[Index];
    }

    return Crc;
}

int main(int argc, char **argv)
{
    ssize_t readSize;
    off_t   skipSize = 0;
    ssize_t fileSize = 0;
    uint32  fileCRC  = 0;
    int     fd;
    char    buffer[100];
    off_t   offsetReturn = 0;

    /* check for valid input */
    if ((argc != 2) || (strncmp(argv[1], "--help", 100) == 0))
    {
        printf("%s\n", CFE_TS_CRC_VERSION_STRING);
        printf("\nUsage: cfe_ts_crc [filename]\n");
        exit(1);
    }
    /* Set to skip the header (116 bytes) */
    skipSize = sizeof(CFE_FS_Header_t) + sizeof(CFE_TBL_File_Hdr_t);

    /* open the input file if possible */
    fd = open(argv[1], O_RDONLY);
    if (fd < 0)
    {
        printf("\ncfe_ts_crc error: can't open input file!\n");
        perror(argv[1]);
        exit(1);
    }
    /* seek past the number of bytes requested */
    offsetReturn = lseek(fd, skipSize, SEEK_SET);
    if (offsetReturn != skipSize)
    {
        printf("\ncfe_ts_crc error: lseek failed!\n");
        printf("%s\n", strerror(errno));
        exit(1);
    }

    /* read the input file 100 bytes at a time */
    do
    {
        readSize = read(fd, buffer, sizeof(buffer));
        if (readSize < 0)
        {
            printf("\ncfe_ts_crc error: file read failed!\n");
            printf("%s\n", strerror(errno));
            exit(1);
        }
        fileCRC = CalculateCRC(buffer, readSize, fileCRC);
        fileSize += readSize;
    } while (readSize > 0);

    /* print the size/CRC results */
    printf("\nTable File Name:            %s\nTable Size:                 %ld Bytes\nExpected TS Validation CRC: "
           "0x%08lX\n\n",
           argv[1], (long)fileSize, (unsigned long)fileCRC);

    /* Close file and check*/
    if (close(fd) != 0)
    {
        printf("\nerror: Cannot close file!\n");
        printf("%s\n", strerror(errno));
        exit(1);
    }

    return 0;
}


cfe_ts_crc_version.h
/************************************************************************
 * NASA Docket No. GSC-18,719-1, and identified as “core Flight System: Bootes”
 *
 * Copyright (c) 2020 United States Government as represented by the
 * Administrator of the National Aeronautics and Space Administration.
 * All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ************************************************************************/

/*! @file
 * @brief Purpose:
 *  @details Provide version identifiers for the ELF to cFE Table Converter. @n
 *  See @ref cfsversions for version and build number and description
 *
 */

#ifndef CFE_TS_CRC_VERSION_H
#define CFE_TS_CRC_VERSION_H

/*
 * Development Build Macro Definitions
 */
#define CFE_TS_CRC_BUILD_NUMBER 24 /*!< @brief Number of commits since baseline */
#define CFE_TS_CRC_BUILD_BASELINE \
    "v1.3.0-rc4+dev" /*!< @brief Development Build: git tag that is the base for the current */

/*
 * Version Macros, see \ref cfsversions for definitions.
 */
#define CFE_TS_CRC_MAJOR_VERSION 1  /*!< @brief Major version number */
#define CFE_TS_CRC_MINOR_VERSION 1  /*!< @brief Minor version number */
#define CFE_TS_CRC_REVISION      99 /*!< @brief Revision version number. Value of 99 indicates a development version.*/

/*!
 * @brief Mission revision.
 *
 * Reserved for mission use to denote patches/customizations as needed.
 * Values 1-254 are reserved for mission use to denote patches/customizations as needed. NOTE: Reserving 0 and 0xFF for
 * cFS open-source development use (pending resolution of nasa/cFS#440)
 */
#define CFE_TS_CRC_MISSION_REV 0xFF

/*
 * Tools to construct version string
 */
#define CFE_TS_CRC_STR_HELPER(x) #x /*!< @brief Helper function to concatenate strings from integer macros */
#define CFE_TS_CRC_STR(x) \
    CFE_TS_CRC_STR_HELPER(x) /*!< @brief Helper function to concatenate strings from integer macros */

/*! @brief Development Build Version Number.
 * @details Baseline git tag + Number of commits since baseline. @n
 * See @ref cfsversions for format differences between development and release versions.
 */
#define CFE_TS_CRC_VERSION CFE_TS_CRC_BUILD_BASELINE CFE_TS_CRC_STR(CFE_TS_CRC_BUILD_NUMBER)

/*! @brief Development Build Version String.
 * @details Reports the current development build's baseline, number, and name. Also includes a note about the latest
 * official version. @n See @ref cfsversions for format differences between development and release versions.
 */
#define CFE_TS_CRC_VERSION_STRING            \
    " cFE TS CRC calculator (tblCRCtool) \n" \
    " DEVELOPMENT BUILD \n"                  \
    " " CFE_TS_CRC_VERSION " \n"             \
    " Last Official Release: tblCRCtool v3.1.0" /* For full support please use official release version */

#endif /* CFE_TS_CRC_VERSION_H */
